File: zad4.cpp
#include <iostream>
#include <time.h>
#include <cmath>
#include <vector>

using namespace std;
// нахождение НОД(a, b) для проверки, являются ли числа взаимопростыми
int VzaimProst(int a, int b) { 

    if (b == 0) return a;

    return VzaimProst(b, a % b);
}
// проверка, является ли число простым
bool CheckPrime(int p) { 

    if (p <= 1) return false;
    for (int i = 2; i * i <= p; ++i) 
    {
        if (p % i == 0) return false;
    }

    return true;
}
 // функция Эйлера
int fi(int num) {

    int result = 1;
    int stepen = 0;

    for (int i = 2; i <= num; ++i)
    {
        while (num % i == 0) 
        {
            stepen++;
            num /= i;
        }
        if (stepen != 0) {
            result *= pow(i, stepen - 1) * (i - 1);
        }
        stepen = 0;
    }

    return result;
}
// нахождение остатка от числа в степени
int Pow(int a, int x, int p) 
{ 

    int result = 1;

    if (CheckPrime(p) == true) 
    { // простое
        x %= p - 1;
    }
    else { // не простое
        x %= fi(p);
    }
    for (int i = 1; i <= x; ++i) {
        result = (result * a) % p;
    }

    return result;
}
// Алгоритм нахождения числа d(c^-1 mod m)
int inversia(int c, int m) 
{ 

    int q;
    vector<int> one = { m, 0 };
    vector<int> two = { c, 1 };
    vector<int> three(2);
    while (two[0] != 0) 
    {
        q = one[0] / two[0];
        three[0] = one[0] % two[0];
        three[1] = one[1] - (q * two[1]);
        one[0] = two[0]; one[1] = two[1];
        two[0] = three[0]; two[1] = three[1];
    }

    if (one[1] < 0) 
    {
        one[1] += m;
    }

    return one[1];
}


int main() {
    setlocale(LC_ALL, "Rus");
    srand(time(0));

    string message;
    cout << "Введите сообщение: ";
    cin >> message;
    int g, n;
    cout << "Введите числа g и n: ";
    cin >> g >> n;

    int x = 100 + rand() % 100000; // абонент A выбирает случайное больше число и генерирует закрытый ключ
    cout << "x = " << x << endl;
    int k = Pow(g, x, n);
    cout << "k = " << k << endl;

    int y = 100 + rand() % 100000; // абонент B выбирает случайное больше число
    while (VzaimProst(y, n - 1) != 1) 
    {
        y = 100 + rand() % 100000;
    }
    cout << "y = " << y << endl;

    int Y = Pow(g, y, n); // абонент B посылает абоненту A
    cout << "Y = " << Y << endl;

    int X = Pow(Y, x, n); // абонент А посылает абоненту B
    cout << "X = " << X << endl;

    int z = inversia(y, n - 1); // абонент B вычисляет
    cout << "z = " << z << endl;
    int ksh = Pow(X, z, n);
    cout << "k' = " << ksh << endl;

    if (k == ksh) 
    {
        int key = k;

        for (int i = 0; i < message.length(); ++i) 
        {
            message[i] = message[i] ^ key;
        }
        cout << "Зашифрованное сообщение: " << message << endl;

        for (int i = 0; i < message.length(); ++i) 
        {
            message[i] = message[i] ^ key;
        }
        cout << "Расшифрованное сообщение: " << message;

    }
    else 
    {
        cout << "Ключи не равны!";
    }

    return 0;
}

File: zad4.py
import random

def VzaimProst(a, b): # нахождение НОД(a, b) для проверки, являются ли числа взаимопростыми
    
    if (b == 0): return a
    
    return VzaimProst(b, a % b)

def CheckPrime(p): # проверка, является ли число простым
    
    if (p <= 1): return False
    for i in range(2, int(p ** 0.5) + 1):
        if (p % i == 0): return False
        
    return True

def fi(num): # функция Эйлера
    
    result = 1
    stepen = 0
    
    for i in range(2, num+1):
        while (num % i == 0):
            stepen += 1
            num //= i
        if (stepen != 0): 
            result *= pow(i, stepen - 1) * (i - 1)
        stepen = 0
        
    return result

def Pow(a, x, p): # нахождение остатка от числа в степени
    
    result = 1
    
    if (CheckPrime(p) == True): x %= p - 1 # простое
    else: x %= fi(p) # не простое
    for i in range(1, x+1):
        result = (result * a) % p
        
    return result

def inversia(c, m): # алгоритм нахождения числа d(c^-1 mod m)
    
    q = 0
    one = [m, 0]
    two = [c, 1]
    three = [0, 0]
    
    while (two[0] != 0):
        q = one[0] // two[0]
        three[0] = one[0] % two[0]
        three[1] = one[1] - (q * two[1])
        one[0], one[1] = two[0], two[1]
        two[0], two[1] = three[0], three[1]
    
    if (one[1] < 0):
        one[1] += m
    
    return one[1]


message = input("Введите сообщение: ")
g = int(input("Введите число g: "))
n = int(input("Введите число n: "))

x = random.randrange(10000, 100000)  # абонент A выбирает случайное больше число и генерирует закрытый ключ
print("x =", x)

k = Pow(g, x, n)
print("k =", k)

y = random.randrange(10000, 100000)  # абонент B выбирает случайное больше число
while VzaimProst(y, n-1) != 1:
    y = random.randrange(10000, 100000)
print("y =", y)

Y = Pow(g, y, n)  # абонент B посылает абоненту A
print("Y =", Y)

X = Pow(Y, x, n)  # абонент А посылает абоненту B
print("X =", X)

z = inversia(y, n-1)  # абонент B вычисляет
print("z =", z)

ksh = Pow(X, z, n)
print("k' =", ksh)

if (k == ksh):
    key = k
    ArrEncrypted = []
    ArrDecrypted = []
    Encrypted = ""
    Decrypted = ""
    
    for i in range(len(message)): # Шифрование
        ArrEncrypted.append(chr(ord(message[i]) ^ key))
    for i in range(len(ArrEncrypted)):
        Encrypted += ArrEncrypted[i]
    print("Зашифрованное сообщение:", Encrypted)
    
    for i in range(len(Encrypted)): # Дешифрование 
        ArrDecrypted.append(chr(ord(Encrypted[i]) ^ key))
    for i in range(len(ArrDecrypted)):
        Decrypted += ArrDecrypted[i]
    print("Расшифрованное сообщение:", Decrypted)
    
else: print("Ключи не равны!")
